File: Mascot Parser Files/csharp_vs2012/example_csharp/create_mgf.cs
/*
##############################################################################
# file: create_mgf.cs                                                        #
# 'msparser' toolkit example code                                            #
##############################################################################
# COPYRIGHT NOTICE                                                           #
# Copyright 1998-2015 Matrix Science Limited  All Rights Reserved.           #
#                                                                            #
##############################################################################
#    $Source: /vol/cvsroot/parser/examples/test_csharp/create_mgf.cs,v $                                #
#     $Author: patricke $                                                    #
#       $Date: 2015/07/31 14:38:19 $                                         #
#   $Revision: 1.2 $                                                         #
# $NoKeywords::                                                            $ #
##############################################################################
*/

using System;
using System.IO;
using matrix_science.msparser;

namespace MsParserExamples
{
    public class create_mgf
    {
        public static void Main(string[] argv)
        {
            if (argv.Length < 1)
            {
                usage();
                return;
            }
            createMGF(argv[0]);
        }

        private static void createMGF(string filename)
        {
            ms_mascotresfile resfile = new ms_mascotresfile(filename);
            if (!resfile.isValid())
            {
                Console.WriteLine("Cannot open results file {0}: {1}", filename, resfile.getLastErrorString());
                return;
            }            

            string outputFilename = filename + ".mgf";
            // check if the mgf file already exists - do not overwrite if it does
            if (File.Exists(outputFilename))
            {
                Console.WriteLine("{0} already exists; will not overwrite", outputFilename);
                return;
            }

            // create the file
            

            using (TextWriter tw = new StreamWriter(outputFilename))
            {
                for (int q = 1; q <= resfile.getNumQueries(); q++)
                {
                    ms_inputquery inpQuery = new ms_inputquery(resfile, q);
                    if (inpQuery.getNumberOfPeaks(1) == 0)
                    {
                        // PMF - just the mass
                        tw.WriteLine(resfile.getObservedMass(q));
                        continue;
                    }

                    tw.WriteLine("BEGIN IONS");
                    tw.WriteLine("PEPMASS={0}", resfile.getObservedMass(q));

                    if (Math.Abs(resfile.getObservedCharge(q)) > 0)
                    {
                        tw.WriteLine("CHARGE={0}", resfile.getObservedCharge(q));
                    }
                    else
                    {
                        tw.WriteLine("CHARGE=Mr");
                    }

                    if (inpQuery.getStringTitle(true).Length > 0)
                    {
                        tw.WriteLine("TITLE={0}", inpQuery.getStringTitle(true));
                    }

                    for (int i = 1; i <= inpQuery.getNumberOfPeaks(1); i++)
                    {
                        tw.WriteLine("{0} {1}", inpQuery.getPeakMass(1, i), inpQuery.getPeakIntensity(1, i));
                    }

                    tw.WriteLine("END IONS");
                }
            }

        }

        private static void usage()
        {
            Console.WriteLine(@"Usage: create_mgf <results file>.

Given a mascot results file name, create an MGF file.  The MGF file
will be named <results file>.mgf in the same directory where <results file>
is located.");
        }
    }
}

File: Python/Mascot/createMGF.py
import msparser
import sys
import os.path

def main( filepath ) :

    createMGF( filepath )


def createMGF(filename) :
    resfile = msparser.ms_mascotresfile(filename)

    if not resfile.isValid() :
        print("Cannot open results file %s : %s" % (filename, resfile.getLastErrorString()))
        return
    
    output_filename = filename + '.mgf'

    if os.path.isfile(output_filename) :
        print("%s already exists; will not overwrite" % output_filename)
        return
    
    try :
        fh = open(output_filename, "w") 
        
    except IOError as err:
        errno, strerror = err.args
        print("Cannot open output_filename for writing: I/O error(%d): %s" % (
                errno, strerror
                ))
        sys.exit(1)

    for q in range(1, 1 + resfile.getNumQueries()) :
        inp_query = msparser.ms_inputquery(resfile, q)

        if inp_query.getNumberOfPeaks(1) == 0 :
            # PMF - just the mass
            fh.write("%f\n" % resfile.getObservedMass(q))
            continue

        fh.write("BEGIN_IONS\n");
        fh.write("PEPMASS=%f\n" % resfile.getObservedMass(q))

        if resfile.getObservedCharge(q) > 0 :
            fh.write("CHARGE=%f+\n" % resfile.getObservedCharge(q))
        else :
            fh.write("CHARGE=Mr\n")

        if inp_query.getStringTitle(1) :
            fh.write("TITLE=%s\n" % inp_query.getStringTitle(1))
        
        for i in range(1, 1 + inp_query.getNumberOfPeaks(1)) :
            fh.write("%f %f\n" % (inp_query.getPeakMass(1, i), inp_query.getPeakIntensity(1, i)))
        
        fh.write("END IONS\n")

    fh.close()
    

def usage() :
    print("""
Usage: create_mgf.py <results file>

Given a mascot results file name, create an MGF file. The MGF file
will be named <results file>.mgf in the same directory where <results file>
is located.
""")

if __name__ == "__main__" :
    sys.exit(main("data.dat"))